sd/internal/instance/cache.go

```diff
@@ -52,10 +54,12 @@ func (c *Cache) Stop() {}
 // Register implements Instancer.
 func (c *Cache) Register(ch chan<- sd.Event) {
 	c.mtx.Lock()
-	defer c.mtx.Unlock()
 	c.reg.register(ch)
+	event := c.state
+	c.mtx.Unlock()
+	eventCopy := copyEvent(event)
 	// always push the current state to new channels
-	ch <- c.state
+	ch <- eventCopy
```

Comment:
I think it's better to put this in Lock. Otherwise the receiver may get stale intances.

Reply:
I don’t think a mutex is solving that problem. You would need updates to arrive in such a way that there would be contention on the mutex. If that is the case, it is probably just as likely you’d have updates arrive in quick succession but miss contention.

Reply:
Lock can ensure that events enter `chan` in order, which means  consumers can consume in the right order.
Make a simple path in your case.
Thread1: Lock -> A event -> UnLock
Thread2: Wait ->         ...                      -> Lock -> B event(newer than A) -> UnLock
Once out of Lock, consumers may consume B first and then A, no one can determine the order.

Reply:
I see your point now. Will fix.